core/src/main/java/com/linecorp/armeria/common/stream/DefaultStreamMessage.java

```diff
@@ -339,38 +346,60 @@ private void notifySubscriber0() {
         }
 
         for (;;) {
-            if (state == State.CLEANUP) {
-                cleanup();
+            if (!beginNotification()) {
                 return;
             }
 
-            final Object o = queue.peek();
-            if (o == null) {
-                break;
-            }
+            for (;;) {
+                if (state == State.CLEANUP) {
+                    cleanup();
+                    return;
+                }
 
-            if (o instanceof CloseEvent) {
-                notifySubscriberWithCloseEvent(subscription, (CloseEvent) o);
-                break;
-            }
+                final Object o = queue.peek();
+                if (o == null) {
+                    break;
+                }
 
-            if (o instanceof AwaitDemandFuture) {
-                if (notifyAwaitDemandFuture()) {
-                    // Notified successfully.
-                    continue;
-                } else {
+                if (o instanceof CloseEvent) {
+                    notifySubscriberWithCloseEvent(subscription, (CloseEvent) o);
+                    break;
+                }
+
+                if (o instanceof AwaitDemandFuture) {
+                    if (notifyAwaitDemandFuture()) {
+                        // Notified successfully.
+                        continue;
+                    } else {
+                        // Not enough demand.
+                        break;
+                    }
+                }
+
+                if (!notifySubscriberWithElements(subscription)) {
                     // Not enough demand.
                     break;
                 }
             }
 
-            if (!notifySubscriberWithElements(subscription)) {
-                // Not enough demand.
-                break;
+            currentThread = null;
+            if (queue.isEmpty() || demand == 0) {
+                return;
             }
         }
     }
 
+    private boolean beginNotification() {
+        if (currentThread == null) {
```

Comment:
I haven't seen this approach before - is it a better option than `ReentrantLock`?

Reply:
I think it's better because the writer does not block. But it has a chance of starvation if the writer keeps writing between the end of the notification loop and `currentThread = null;`.

Reply:
I used the `currentThread` as just a flag to get a permission(?) to write.
I think it's better than using a lock. This is the jmh result.

Before adding guard
Benchmark                                                    (num)   Mode  Cnt        Score       Error  Units
StreamMessageBenchmark.defaultStreamMessage_jmhEventLoop         3  thrpt   20  1819437.496 ± 85424.992  ops/s
StreamMessageBenchmark.defaultStreamMessage_jmhEventLoop         5  thrpt   20  1306154.513 ± 11078.302  ops/s
StreamMessageBenchmark.defaultStreamMessage_jmhEventLoop        20  thrpt   20   408677.782 ±  3581.340  ops/s
StreamMessageBenchmark.defaultStreamMessage_jmhEventLoop       100  thrpt   20    88740.796 ±   718.324  ops/s
StreamMessageBenchmark.defaultStreamMessage_jmhEventLoop      1000  thrpt   20     8452.984 ±   100.549  ops/s
StreamMessageBenchmark.defaultStreamMessage_noExecutor           3  thrpt   20  1904230.990 ± 21922.003  ops/s
StreamMessageBenchmark.defaultStreamMessage_noExecutor           5  thrpt   20  1352462.653 ± 12377.573  ops/s
StreamMessageBenchmark.defaultStreamMessage_noExecutor          20  thrpt   20   410967.580 ±  3301.210  ops/s
StreamMessageBenchmark.defaultStreamMessage_noExecutor         100  thrpt   20    87021.066 ±   845.651  ops/s
StreamMessageBenchmark.defaultStreamMessage_noExecutor        1000  thrpt   20     8797.847 ±   113.686  ops/s
StreamMessageBenchmark.defaultStreamMessage_notJmhEventLoop      3  thrpt   20   109138.428 ±  3071.426  ops/s
StreamMessageBenchmark.defaultStreamMessage_notJmhEventLoop      5  thrpt   20   107635.186 ±   655.422  ops/s
StreamMessageBenchmark.defaultStreamMessage_notJmhEventLoop     20  thrpt   20    86603.677 ±  4748.522  ops/s
StreamMessageBenchmark.defaultStreamMessage_notJmhEventLoop    100  thrpt   20    47380.294 ±  5681.907  ops/s
StreamMessageBenchmark.defaultStreamMessage_notJmhEventLoop   1000  thrpt   20     8355.908 ±    46.945  ops/s


After adding guard
Benchmark                                                    (num)   Mode  Cnt        Score       Error  Units
StreamMessageBenchmark.defaultStreamMessage_jmhEventLoop         3  thrpt   20  1802002.636 ± 20998.705  ops/s
StreamMessageBenchmark.defaultStreamMessage_jmhEventLoop         5  thrpt   20  1223637.555 ± 24717.232  ops/s
StreamMessageBenchmark.defaultStreamMessage_jmhEventLoop        20  thrpt   20   367987.533 ±  4001.446  ops/s
StreamMessageBenchmark.defaultStreamMessage_jmhEventLoop       100  thrpt   20    78079.797 ±  1648.554  ops/s
StreamMessageBenchmark.defaultStreamMessage_jmhEventLoop      1000  thrpt   20     8232.707 ±   120.826  ops/s
StreamMessageBenchmark.defaultStreamMessage_noExecutor           3  thrpt   20  1811347.170 ± 26448.754  ops/s
StreamMessageBenchmark.defaultStreamMessage_noExecutor           5  thrpt   20  1254014.492 ± 11579.483  ops/s
StreamMessageBenchmark.defaultStreamMessage_noExecutor          20  thrpt   20   377817.083 ±  4845.750  ops/s
StreamMessageBenchmark.defaultStreamMessage_noExecutor         100  thrpt   20    81239.262 ±   910.667  ops/s
StreamMessageBenchmark.defaultStreamMessage_noExecutor        1000  thrpt   20     8114.463 ±   108.310  ops/s
StreamMessageBenchmark.defaultStreamMessage_notJmhEventLoop      3  thrpt   20   108617.160 ±  2045.207  ops/s
StreamMessageBenchmark.defaultStreamMessage_notJmhEventLoop      5  thrpt   20   106915.884 ±   868.068  ops/s
StreamMessageBenchmark.defaultStreamMessage_notJmhEventLoop     20  thrpt   20    89596.127 ±  1207.321  ops/s
StreamMessageBenchmark.defaultStreamMessage_notJmhEventLoop    100  thrpt   20    47020.757 ±  1196.503  ops/s
StreamMessageBenchmark.defaultStreamMessage_notJmhEventLoop   1000  thrpt   20     7604.880 ±    93.333  ops/s

Reply:
Won't the writer not block if using `ReentrantLock.tryLock`? Looking at code of `ReentrantLock` and `NonfairSync` it seems very similar to this code, so it might result in simpler code to use it, but not a big deal either way.